Lib/concurrent/futures/process.py

```diff
@@ -400,12 +411,14 @@ def process_result_item(self, result_item):
         if isinstance(result_item, int):
             # Clean shutdown of a worker using its PID
             # (avoids marking the executor broken)
-            assert self.is_shutting_down()
             p = self.processes.pop(result_item)
             p.join()
-            if not self.processes:
-                self.join_executor_internals()
-                return
+            if self.is_shutting_down() and not self.pending_work_items:
+                if not self.processes:
+                    self.join_executor_internals()
+            else:
+                if executor := self.executor_reference():
+                    executor._add_process_to_pool()
```

Comment:
Two questions:
- should it take the `_shutdown_lock`?
- why not call `_adjust_process_count`? if some workers are idle we may not want to spawn a new one immediately

Reply:
Hi @pitrou thanks for the review.

> Should it take the `_shutdown_lock`

I'm really not sure. What would you imagine this doing? I can add the shutdown_lock wherever you think it makes sense, but it's unclear to me where this should go

> why not call _adjust_process_count? if some workers are idle we may not want to spawn a new one immediately

You may have stumbled on to a larger problem here, but I'll try to explain why I'm forcefully spawning a new process. The problem is that the semaphore can sort of lie. Once a worker finishes it's last task, it exits by first putting it's `ResultItem` on the queue, and *then* putting the pid on the queue.

So functionally what happens is that the `ResultItem` gets processed, and the semaphore gets released which indicates that we have a process available for use. The manager thread then receives the next "result item" which is actually the worker that just exited cleanly, so if I called `_adjust_process_count` we are guaranteed to have an "idle" worker which is actually just dead.

Hopefully that answers the question.

Reply:
After thinking about the semaphore problem a little bit more, it occurred to me that perhaps a more elegant way of solving this problem is by modifying the `ResultItem` and allowing it to indicate that a clean shutdown of the process occurred. This would allow us to keep the previous assumption that if a `ResultItem` was an integer, that a process exited and something went wrong.

Not sure if this is a good solution or not, but it seems good on the surface. @pitrou thoughts?